message/infrastructure/googlecloud/subscriber.go

```diff
@@ -0,0 +1,221 @@
+	config.setDefaults()
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid config")
+	}
+
+	client, err := pubsub.NewClient(ctx, config.ProjectID, config.ClientOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &subscriber{
+		closing: make(chan struct{}, 1),
+		closed:  false,
+
+		allSubscriptionsWaitGroup: sync.WaitGroup{},
+		activeSubscriptions:       map[string]struct{}{},
+		activeSubscriptionsLock:   sync.RWMutex{},
+
+		client: client,
+		config: config,
+
+		unmarshaler: config.Unmarshaler,
+		logger:      logger,
+	}, nil
+}
+
+func (s *subscriber) Subscribe(topic string) (chan *message.Message, error) {
+	// todo: pass root ctx from somewhere?
+	ctx, cancel := context.WithCancel(context.Background())
+
+	if s.closed {
+		return nil, ErrSubscriberClosed
+	}
+
+	logFields := watermill.LogFields{
+		"provider":          ProviderName,
+		"topic":             topic,
+		"consumers_count":   s.config.ConsumersCount,
+		"subscription_name": s.config.SubscriptionName,
+	}
+	s.logger.Info("Subscribing to Google Cloud PubSub topic", logFields)
+
+	output := make(chan *message.Message, 0)
+
+	s.allSubscriptionsWaitGroup.Add(1)
+
+	sub, err := s.subscription(ctx, topic)
+	if err != nil {
+		return nil, err
+	}
+
+	err = sub.Receive(ctx, func(ctx context.Context, pubsubMsg *pubsub.Message) {
+		msg, err := s.unmarshaler.Unmarshal(pubsubMsg)
+		if err != nil {
+			s.logger.Error("could not unmarshal Google Cloud PubSub message", err, nil)
+			return
+		}
+
+		output <- msg
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		<-s.closing
+		s.logger.Debug("Closing message consumer", logFields)
+		cancel()
+
+		close(output)
+		s.allSubscriptionsWaitGroup.Done()
+	}()
+
+	return output, nil
+}
+
+func (s *subscriber) Close() error {
+	if s.closed {
+		return nil
+	}
+
+	s.closed = true
+	close(s.closing)
+	s.allSubscriptionsWaitGroup.Wait()
+
+	err := s.client.Close()
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debug("Google Cloud PubSub subscriber closed", nil)
+	return nil
+}
+
+const subscriptionIDTemplate = "watermill_%s_%d"
+
+// subscription obtains a subscription object. If subscription doesn't exist on PubSub, create it.
+// subsequent calls to `subscription` with the same `topic` return separate subscriptions,
+// with ids according to `subscriptionIDTemplate`.
+func (s *subscriber) subscription(ctx context.Context, topic string) (sub *pubsub.Subscription, err error) {
+	s.activeSubscriptionsLock.RLock()
+	var subscriptionID string
+	for i := 0; ; i++ {
+		subscriptionID = fmt.Sprintf(subscriptionIDTemplate, topic, i)
+		if _, ok := s.activeSubscriptions[subscriptionID]; ok {
+			continue
+		}
+	}
+	s.activeSubscriptionsLock.RUnlock()
+
+	defer func() {
+		if err != nil {
+			s.activeSubscriptionsLock.Lock()
+			s.activeSubscriptions[sub.ID()] = struct{}{}
+			s.activeSubscriptionsLock.Unlock()
+		}
+	}()
+
+	sub = s.client.Subscription(subscriptionID)
```

Comment:
isn't subscriptionID something like consumer group?

Reply:
if yes, it should be probably injected to the constructor

Reply:
It is somewhat similar. See 532ef915c1ca7188e48f9639082219c9c7d2054b, now it is taken from the config. I will add validation to make sure that it is not left empty.